Add a "polling log" page as a link on the project page

proski opened the pull request:
```
*  Add a "Polling Log" page as a link on the project page
   
   Add StashPollingAction class, which is an Action and a logger. Use slf4j
   MessageFormatter for the log messages.
   
   StashPollingActionFactory is used to inject StashPollingAction into the
   project using the Extension mechanism.
   
   Change StashBuildTrigger and StashRepository to write information about
   polling to the polling log. Some messages are now logged both to the
   polling log and to the global Jenkins log.
```

Without this change, the plugin writes to the Jenkins log for every poll. That makes it a major source of log messages that provide very little value. 

In most cases, the Jenkins administrator doesn't care when the polls happened in the past. To check if polling is working, it's sufficient to check the time of the last poll.

Jenkins administrators may also be interested why a particular pull request is not triggering a build. The latest reason not to build the PR should be sufficient. It's not like the reason would change over time, and even if it changes, the last reason is most relevant.

Other triggers, such as Git Plugin, provide a polling log link on the project page. This PR makes Stash Pull Request Builder do the same.

This PR takes care of the infrastructure. The polling log contains the time when the last poll was run, its duration and the number of PRs to be built. More logging can be moved to the poll log in the future.

proski commented:
Reimplemented using `PrintWriter` to avoid dealing with FindBugs warnings and `UnsupportedEncodingException`. If only `PrintStream` allowed me to pass `StandardCharsets.UTF_8` to its constructor, but no, they made that constructor private!

We could use a real logger eventually. The plugin has a transitive dependency on slf4j, it could be made a direct dependency.

It looks like `getProjectActions()` is not flexible enough to prevent broken links on matrix projects. `TransientActionFactory` should be more powerful, I'm going to try it.

By the way, git plugin has the same broken links on matrix projects, and it's using `TransientActionFactory` though SCM API, but maybe it's not using it correctly.

proski committed:
Add a "Polling Log" page as a link on the project page

Add StashPollingAction class, which is an Action and a logger. Use slf4j
MessageFormatter for the log messages. Add slf4j dependency.

StashPollingActionFactory is used to inject StashPollingAction into the
project using the Extension mechanism.

Change StashBuildTrigger and StashRepository to write information about
polling to the polling log. Some messages are now logged both to the
polling log and to the global Jenkins log.

proski commented:
Converted to `TransientActionFactory`, it's the modern approach: https://wiki.jenkins.io/display/JENKINS/Action+and+its+family+of+subtypes

It turns out the unsightly broken links on non-root matrix project pages exist even without this PR. I should have checked that. It must be some other plugin. Perhaps for the first time, Stash Pull Request Builder is not the worst plugin in my Jenkins setup!

Since `StashPollingAction` doesn't throw anymore, I can use it to collect logs from the `StashBuildTrigger` class, which I started doing for "Can't start trigger". That could eventually include such information as invalid credential ID (not in this PR).

The only remaining work for this PR is adding unit tests.

The silence in the logs is amazing!

jakub-bochenski commented on src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepositoryTest.java:
```diff
@@ -129,6 +129,10 @@ private void jobSetup(ParameterDefinition... parameterDefinitions) {
         .thenReturn(Arrays.asList(parameterDefinitions));
   }
 
+  private void triggerSetup() {
```
that's a lot of repetition, can you make it a `@Before` method?

proski replied:
Done. Don't know why it didn't occur to me :smile:

proski commented:
Fixed double escaping of the log text. Jelly escapes the text already when `escape-by-default='true'` is used. Added unit tests for the new classes with 100% code coverage.

proski commented:
Using slg4j `MessageFormatter` for the `StashPollingAction#log` arguments. Since we are adding a new logger, we can use the most advanced formatter. It's good to have `{}` for all arguments without any risk to misplace `%d` and `%f`. Also, the exceptions are recognized at the end on argument lists and preserved for extra processing.

Since jenkins depends on slf4j, no extra jars are bundled with the plugin.
